AMDGPU: Add tonearest and towardzero roundings for intrinsic llvm.fptrunc.round

changpeng opened the pull request:
This work simplifies and generalizes the instruction definition for intrinsic llvm.fptrunc.round. We no longer name the instruction with the rounding mode. Instead, we introduce an immediate operand for the rounding mode for the pseudo instruction. This immediate will be used to set up the hardware mode register at the time the real instruction is generated. We name the pseudo instruction as FPTRUNC_ROUND_F16_F32 (for f32 -> f16), which is easy to generalize for other types.

  "round.towardzero" and "round.tonearest" are added for f32 -> f16 truncating, in addition to the existing "round.upward" and "round.downward". Other rounding modes are not supported by hardware at this moment.

changpeng committed:
AMDGPU: Improve codegen for intrinsic llvm.fptrunc.round

  This work simplifies and generalizes the intruction definition and
code generation for intrinsic llvm.fptrunc.round. We no longer name
the instruction with the rounding mode. Instead, we introduce an
immediate operand for the rounding mode for the pseudo instruction.
This immediate will be used to set up the hardware mode register at
the time the real instruction is generated. We name the pseudo instruction
as FPTRUNC_ROUND_F32_TO_F16 (for f32 -> f16), which is easy to generalize for
other types.
  "round.towardzero" and "round.tonearest" are added for f32 -> f16
truncating, in addition to the existing "round.upward" and "round.downward".
Other rounding modes are not supported by hardware at this moment.
  We can easily extend this work to support f64 -> f32 and f64 -> f16,
which will be the patch that follows. Support of truncating to bfloat
is also in the plan.

jayfoad commented:
> Improve codegen for intrinsic llvm.fptrunc.round

Do you have an example where codegen is improved?

> We name the pseudo instruction as FPTRUNC_ROUND_F32_TO_F16 (for f32 -> f16)

How about calling it `FPTRUNC_ROUND_F16_F32` to match `V_CVT_F16_F32`?

arsenm commented on llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:
```diff
@@ -3,6 +3,15 @@
+; CHECK:       ; %bb.0:
+; CHECK-NEXT:    v_cvt_f16_f32_e32 v0, v0
+; CHECK-NEXT:    ; return to shader part epilog
+  %res = call half @llvm.fptrunc.round.f16.f32(float %a, metadata !"round.tonearest")
```
this should work, but hopefully instcombine just turns this into fptrunc?

changpeng replied:
At custom lowering the the intrinsic, we can turn it to fp_round, which can also generated the expected output for "round.tonearest". However, it seems the lowering of fp_round ignores the rounding mode operand in most cases (just assume default of round-to-nearest-even)

arsenm replied:
that shouldn't be custom lowering, that should be a generic lowering (and instcombine optimization)

changpeng replied:
I would put the instcombine optimization for llvm.fptrunc.round of mode round.tonearest to a separate patch. Thanks.

changpeng replied:
Can we do something in instcombine for round.towardzero also?

arsenm replied:
> I would put the instcombine optimization for llvm.fptrunc

Yes

> Can we do something in instcombine for round.towardzero also?

I assume quite a few combines could be done with it, but you can't merely replace it with anything obvious

changpeng replied:
For f32 -> bf16, you won't be able to rtz for v_cvt_pk_bf16_f32 (hardware uses round to nearest even all the times).
So if you really want to rtz, you will have to do in the compiler.

arsenm replied:
Yes, it will need some kind of expansion

arsenm reviewed (commented):
Title is misleading, this doesn't change codegen at all?

changpeng commented:
> Title is misleading, this doesn't change codegen at all?
The patch changes the instruction definition, and adds the cases for two additional rounding modes. I was thinking, the whole process for intrinsic to generate assembly code is called code generation. But codegen may have a specific meaning.

changpeng commented:
> Do you have an example where codegen is improved?
Same code generated for round.upward ad round.downward. Add two additional cases round.tonearst and round.towardzero.
 
> How about calling it `FPTRUNC_ROUND_F16_F32` to match `V_CVT_F16_F32`?
Will do as suggested. Thanks.

changpeng committed:
AMDGPU: Add tonearest and towardzero support for llvm.fptrunc.round

  This work simplifies and generalizes the instruction definition and
for intrinsic llvm.fptrunc.round. We no longer name the instruction with
the rounding mode. Instead, we introduce an immediate operand for the
rounding mode for the pseudo instruction. This immediate will be used
to set up the hardware mode register at the time the real instruction is
generated. We name the pseudo instruction as FPTRUNC_ROUND_F16_F32
(for f32 -> f16), which is easy to generalize for other types.

  "round.towardzero" and "round.tonearest" are added for f32 -> f16
truncating, in addition to the existing "round.upward" and "round.downward".
Other rounding modes are not supported by hardware at this moment.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -7089,23 +7089,29 @@ bool AMDGPULegalizerInfo::legalizeBVHIntrinsic(MachineInstr &MI,
-  else
+  // Only support towardzero, tonearest, upward and downward.
+  int RoundMode = MI.getOperand(2).getImm();
+  if (RoundMode < (int)RoundingMode::TowardZero ||
+      RoundMode > (int)RoundingMode::TowardNegative)
```
Do exact equality checks for the supported modes

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6671,10 +6655,34 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG,
+
+  // Only support towardzero, tonearest, upward and downward.
+  int RoundMode = Op.getConstantOperandVal(1);
+  if (RoundMode < (int)RoundingMode::TowardZero ||
+      RoundMode > (int)RoundingMode::TowardNegative)
```
Same, do exact equality checks for the supported modes

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6671,10 +6655,34 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG,
   assert(Op.getValueType() == MVT::f16 &&
          "Do not know how to custom lower FP_ROUND for non-f16 type");
 
+  return Op;
```
Making the rest of the function dead, which must break something

changpeng replied:
Sorry, this was the leftover when I tried something else. should not change fp_round

changpeng committed:
AMDGPU: Add tonearest and towardzero roundings for intrinsic llvm.fptrunc.round

 Update based on comments:
  1. Do exact equality checks for the supported modes.
  2. Remove a bad leftover code in LowerFP_ROUND.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2180,7 +2180,7 @@ bool AMDGPULegalizerInfo::legalizeCustom(
     return legalizeCTLZ_ZERO_UNDEF(MI, MRI, B);
   case TargetOpcode::G_INTRINSIC_FPTRUNC_ROUND:
-    return legalizeFPTruncRound(MI, B);
+    return legalizeFPTruncRound(MI, MRI, B);
```
Don't need to add this parameter, you can just get MRI from the MachineIRBuilder

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -7089,23 +7089,31 @@ bool AMDGPULegalizerInfo::legalizeBVHIntrinsic(MachineInstr &MI,
 
-  B.buildInstr(Opc)
+  // "round.towardzero" -> TowardZero 0        -> FP_ROUND_ROUND_TO_ZERO 3
+  // "round.tonearest"  -> NearestTiesToEven 1 -> FP_ROUND_ROUND_TO_NEAREST 0
+  // "round.upward"     -> TowardPositive 2    -> FP_ROUND_ROUND_TO_INF 1
+  // "round.downward    -> TowardNegative 3    -> FP_ROUND_ROUND_TO_NEGINF 2
+  unsigned HW_Mode = (RoundMode + 3) % 4;
+  B.buildInstr(AMDGPU::G_FPTRUNC_ROUND)
       .addDef(MI.getOperand(0).getReg())
-      .addUse(MI.getOperand(1).getReg());
+      .addUse(Src)
+      .addImm(HW_Mode);
```
All this is really doing is just remapping the encoded rounding mode values. At that point it should just be a selection pattern

changpeng replied:
Do you mean we can use a patch to directly map intrinsic to pseudo? I can try that in a separate patch.

arsenm replied:
Yes

arsenm commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -225,20 +225,12 @@ def S_INVERSE_BALLOT_U64 : SPseudoInstSI<
-def FPTRUNC_DOWNWARD_PSEUDO : VPseudoInstSI <(outs VGPR_32:$vdst),
-  (ins VGPR_32:$src0),
-  [(set f16:$vdst, (SIfptrunc_round_downward f32:$src0))]>;
+def FPTRUNC_ROUND_F16_F32 : VPseudoInstSI <(outs VGPR_32:$vdst),
```
This needs to be more obviously a pseudo from the name

changpeng replied:
FPTRUNC_ROUND_PSEUDO_F16_F32 ?

arsenm replied:
FPTRUNC_ROUND_F16_F32_PSEUDO?

arsenm commented on llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll:
```diff
@@ -23,6 +32,16 @@ define amdgpu_gs half @v_fptrunc_round_f32_to_f16_downward(float %a) {
+  %res = call half @llvm.fptrunc.round.f16.f32(float %a, metadata !"round.towardzero")
+  ret half %res
+}
+
```
can you also include a test where multiple modes are used in the same function

changpeng replied:
Add tests that has multiple calls with round.towardzero and round.downward.

changpeng committed:
AMDGPU: Add tonearest and towardzero roundings for intrinsic llvm.fptrunc.round

  Add muitiple calls tests for round.downward and round.towardzero

changpeng committed:
AMDGPU: Add tonearest and towardzero roundings for intrinsic llvm.fptrunc.round

  Rename the pseudo to FPTRUNC_ROUND_F16_F32_PSEUDO

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6671,6 +6655,30 @@ SDValue SITargetLowering::getFPExtOrFPRound(SelectionDAG &DAG,
+  // "round.downward    -> TowardNegative 3    -> FP_ROUND_ROUND_TO_NEGINF 2
+  unsigned HW_Mode = (RoundMode + 3) % 4;
+  SDLoc DL(Op);
+  unsigned Opc = AMDGPUISD::FPTRUNC_ROUND;
```
can lose the Opc variable (or just move this all to a pattern)

changpeng committed:
AMDGPU: Add tonearest and towardzero roundings for intrinsic llvm.fptrunc.round

  Remove MRI parameter for legalizeFPTruncRound function.

arsenm reviewed (approved):
lgtm but this should still just be a pattern

changpeng commented:
> lgtm but this should still just be a pattern

Do you mean the pattern to map "ISD::FPTRUNC_ROUND src, round" to pseudo?
I am not sure how to translate round to the hardware rounding mode during the pattern matching!

arsenm commented:
> > lgtm but this should still just be a pattern
> 
> Do you mean the pattern to map "ISD::FPTRUNC_ROUND src, round" to pseudo? I am not sure how to translate round to the hardware rounding mode during the pattern matching!

Yes. You just need an SDNodeXForm to modify the value

changpeng commented:
> Yes. You just need an SDNodeXForm to modify the value

What about a rounding mode that is not supported, such as "round.tonearestaway"? 

I am doing like following, but it does not show that the intrinsic can not be selected:

def as_supported_rounding_mode: SDNodeXForm<imm, [{
  int RoundMode = N->getSExtValue();
  if (RoundMode == (int)RoundingMode::TowardZero ||
      RoundMode == (int)RoundingMode::NearestTiesToEven ||
      RoundMode == (int)RoundingMode::TowardPositive ||
      RoundMode == (int)RoundingMode::TowardNegative) {
    unsigned HW_Mode = (RoundMode + 3) % 4;
    return CurDAG->getTargetConstant(HW_Mode, SDLoc(N), MVT::i32);
  }
  return SDValue();
}]>;

changpeng commented:
And what to do on global-isel path? I guess something similar. 
It will end up with similar complexity as before.

arsenm commented:
> And what to do on global-isel path? I guess something similar. It will end up with similar complexity as before.

You use the same pattern, the SDNodeXForm implementation is just different and uses a custom renderer function. The overall thing will be 1/5th the code you have here for the 2x custom lowering

arsenm commented:
> > Yes. You just need an SDNodeXForm to modify the value
> 
> What about a rounding mode that is not supported, such as "round.tonearestaway"?
> 

You can use a TImmLeaf predicate on the immediate value (which can also combine the xform)
